packages/animations/lib/src/dual_transition_builder.dart

```diff
@@ -0,0 +1,228 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+
+/// Builder callback used by [DualTransitionBuilder].
+///
+/// The builder is expected to return a transition powered by the provided
+/// `animation` and wrapping the provided `child`.
+///
+/// The `animation` provided to the builder always runs forward from 0.0 to 1.0.
+/// The builder will be called whenever the [Animation.value] of `animation`
+/// has changed.
+typedef TransitionBuilder = Widget Function(
+  BuildContext context,
+  Animation<double> animation,
+  Widget child,
+);
+
+/// A animated builder that animates its [child] with different transitions
+/// based on the [AnimationStatus] of the provided [animation].
```

Comment:
Maybe we could shorten the first sentence a little, and then provide a little more background on the next paragraph to describe why this solution is necessary as context for future developers:
```
/// A transition builder that animates its [child] based on the [AnimationStatus]
/// of the provided [animation].
///
/// Using this builder allows the widget tree to maintain its shape by nesting the 
/// enter and exit transitions. This prevents animations that start during the 
/// transition animation from resetting upon completion of the animation, which
/// could happen if the enter and exit transitions result in differently-shaped
/// widget trees.
```

The description I provided still seems confusing for those who've never observed the problem before, but I'd like to detail this context as clearly as possible.